tests/test_decoders.py

```diff
@@ -0,0 +1,201 @@
+        assert_called_with_tensors(layer1, [call_1])
+
+    def test_forward_more_than_2_layers(self):
+        """If there are more than 2 layers, they must be called adding data to each result"""
+        # setup
+        layer1 = Mock(return_value=torch.Tensor([[5, 6], [7, 8]]))
+        layer2 = Mock(return_value=torch.Tensor([[9, 10], [11, 12]]))
+        layer3 = Mock(return_value=torch.Tensor([[13, 14], [15, 16]]))
+        self.test_decoder._models = [layer1, layer2, layer3]
+
+        # run
+        image = torch.Tensor([[1, 2], [3, 4]])
+        result = self.test_decoder.forward(image)
+
+        # asserts
+        call_layer_1 = call(torch.Tensor([[1, 2], [3, 4]]))
+        call_layer_2 = call(torch.Tensor([[5, 6], [7, 8]]))
+        call_layer_3 = call(torch.Tensor([[5, 6, 9, 10], [7, 8, 11, 12]]))
+
+        assert_called_with_tensors(layer1, [call_layer_1])
+        assert_called_with_tensors(layer2, [call_layer_2])
+        assert_called_with_tensors(layer3, [call_layer_3])
+
+        assert (result == torch.Tensor([[13, 14], [15, 16]])).all()
+
+
+class TestDenseDecoder(TestCase):
+
+    class TestDecoder(decoders.DenseDecoder):
+        def __init__(self):
+            pass
+
+    def test_upgrade_legacy_without_version(self):
+        """Upgrade legacy must create self._models from conv1, conv2, conv3, conv4"""
+        # setup
+        test_decoder = self.TestDecoder() # instance an empty decoder
+        test_decoder.conv1 = Mock(return_value=torch.Tensor([[5, 6], [7, 8]]))
+        test_decoder.conv2 = Mock(return_value=torch.Tensor([[9, 10], [11, 12]]))
+        test_decoder.conv3 = Mock(return_value=torch.Tensor([[13, 14], [15, 16]]))
+        test_decoder.conv4 = Mock(return_value=torch.Tensor([[17, 18], [19, 20]]))
+
+        expected_models = [
+            test_decoder.conv1,
+            test_decoder.conv2,
+            test_decoder.conv3,
+            test_decoder.conv4,
+        ]
+
+        expected_version = '1'
+
+        # run
+        test_decoder.upgrade_legacy()
+
+        # assert
+        assert test_decoder._models == expected_models
+        assert test_decoder.version == expected_version
+
+    @patch('steganogan.decoders.nn.Sequential', autospec=True)
+    def test_upgrade_legacy_with_version_1(self, sequential_mock):
+        """The object must be the same and not changed by the method"""
+        # setup
+        decoder = decoders.DenseDecoder(1, 1)
+        expected = copy.deepcopy(decoder)
+
+        # run
+        decoder.upgrade_legacy()
+
+        # assert
+        assert decoder.__dict__ == expected.__dict__
+
+
+    @patch('steganogan.decoders.nn.Sequential')
+    @patch('steganogan.decoders.nn.Conv2d')
+    @patch('steganogan.decoders.nn.BatchNorm2d')
+    def test___init__(self, batchnorm_mock, conv2d_mock, sequential_mock):
+        """Test the init params and that the layers are created correctly"""
+        # setup
+        data_depth = 2
```

Comment:
Do not put a "setup" block if there is no "setup action", like creating mocks or complex objects.